DynamicTablesPkg/Library/Acpi/Common/AcpiDbg2Lib/Dbg2Generator.c

```diff
@@ -266,33 +164,160 @@ FreeDbg2TableEx (
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Memory range length too large. Length = %u\n", MemoryRanges[Index].Length));
+      return EFI_INVALID_PARAMETER;
+    }
+  }
+
+  // Calculate total size with overflow check
+  TotalSize = (UINT64)sizeof (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT) +
+              ((UINT64)sizeof (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE) + sizeof (UINT32)) * MemoryRangesCount +
+              (UINT64)sizeof (CHAR8) * DBG2_NAMESPACESTRING_FIELD_SIZE;
+
+  if (TotalSize > MAX_UINT16) {
+    DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Device size too large\n"));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Dbg2DeviceSize    = (UINT16)TotalSize;
+  CurrentDbg2Device = *AcpiDbg2Device;
+
+  // Initialize device structure
+  CurrentDbg2Device->Revision                        = EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT_REVISION;
+  CurrentDbg2Device->Length                          = Dbg2DeviceSize;
+  CurrentDbg2Device->NumberofGenericAddressRegisters = (UINT8)MemoryRangesCount;
+  CurrentDbg2Device->NameSpaceStringLength           = DBG2_NAMESPACESTRING_FIELD_SIZE;
+  CurrentDbg2Device->NameSpaceStringOffset           = sizeof (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT);
+  CurrentDbg2Device->OemDataLength                   = 0;
+  CurrentDbg2Device->OemDataOffset                   = 0;
+  CurrentDbg2Device->PortType                        = DeviceInfo->PortType;
+  CurrentDbg2Device->PortSubtype                     = DeviceInfo->PortSubtype;
+  CurrentDbg2Device->BaseAddressRegisterOffset       = CurrentDbg2Device->NameSpaceStringOffset + DBG2_NAMESPACESTRING_FIELD_SIZE;
+  CurrentDbg2Device->AddressSizeOffset               = CurrentDbg2Device->BaseAddressRegisterOffset +
+                                                       ((UINT16)sizeof (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE) * ((UINT8)MemoryRangesCount));
+  for (Index = 0; Index < MemoryRangesCount; Index++) {
+    BaseAddressRegister                    = (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE *)((UINT8 *)CurrentDbg2Device + CurrentDbg2Device->BaseAddressRegisterOffset + (Index * sizeof (EFI_ACPI_6_3_GENERIC_ADDRESS_STRUCTURE)));
+    AddressSize                            = (UINT32 *)((UINT8 *)CurrentDbg2Device + CurrentDbg2Device->AddressSizeOffset + (Index * sizeof (UINT32)));
+    BaseAddressRegister->AddressSpaceId    = EFI_ACPI_6_3_SYSTEM_MEMORY;
+    BaseAddressRegister->RegisterBitWidth  = 32;
+    BaseAddressRegister->RegisterBitOffset = 0;
+    BaseAddressRegister->AccessSize        = DeviceInfo->AccessSize;
+    BaseAddressRegister->Address           = MemoryRanges[Index].BaseAddress;
+    *AddressSize                           = (UINT32)(MemoryRanges[Index].Length);
+  }
+
+  if (DeviceInfo->ObjectName[0] == '\0') {
+    // If device string is empty then use "." as the name per the DBG2 specification.
+    AsciiSPrint ((CHAR8 *)CurrentDbg2Device + CurrentDbg2Device->NameSpaceStringOffset, DBG2_NAMESPACESTRING_FIELD_SIZE, ".");

```

Comment:
Nit: is it possible to create a variable containing the pointer to `CurrentDbg2Device + CurrentDbg2Device->NameSpaceStringOffset` and add newlines to the function call ? This makes really long lines in the code (the CI doesn't seem to have an issue with that though)